osu.Game/Online/Spectator/SpectatorScoreProcessor.cs

```diff
@@ -0,0 +1,192 @@
+        }
+
+        private void onSpectatorStatesChanged(object? sender, NotifyDictionaryChangedEventArgs<int, SpectatorState> e)
+        {
+            if (!spectatorStates.TryGetValue(userId, out var userState) || userState.BeatmapID == null || userState.RulesetID == null)
+            {
+                scoreProcessor?.RemoveAndDisposeImmediately();
+                scoreProcessor = null;
+                scoreInfo = null;
+                spectatorState = null;
+                replayFrames.Clear();
+                return;
+            }
+
+            if (scoreProcessor != null)
+                return;
+
+            Debug.Assert(scoreInfo == null);
+
+            RulesetInfo? rulesetInfo = rulesetStore.GetRuleset(userState.RulesetID.Value);
+            if (rulesetInfo == null)
+                return;
+
+            Ruleset ruleset = rulesetInfo.CreateInstance();
+
+            spectatorState = userState;
+            scoreInfo = new ScoreInfo { Ruleset = rulesetInfo };
+            scoreProcessor = ruleset.CreateScoreProcessor();
+            scoreProcessor.Mods.Value = userState.Mods.Select(m => m.ToMod(ruleset)).ToArray();
+        }
+
+        private void onNewFrames(int incomingUserId, FrameDataBundle bundle)
+        {
+            if (incomingUserId != userId)
+                return;
+
+            Schedule(() =>
+            {
+                if (scoreProcessor == null)
+                    return;
+
+                replayFrames.Add(new TimedFrame(bundle.Frames.First().Time, bundle.Header));
+                UpdateScore();
+            });
+        }
+
+        public void UpdateScore()
+        {
+            if (scoreInfo == null || replayFrames.Count == 0)
+                return;
+
+            Debug.Assert(spectatorState != null);
+            Debug.Assert(scoreProcessor != null);
+
+            int frameIndex = replayFrames.BinarySearch(new TimedFrame(ReferenceClock.CurrentTime));
+            if (frameIndex < 0)
+                frameIndex = ~frameIndex;
+            frameIndex = Math.Clamp(frameIndex - 1, 0, replayFrames.Count - 1);
+
+            TimedFrame frame = replayFrames[frameIndex];
+            Debug.Assert(frame.Header != null);
+
+            scoreInfo.MaxCombo = frame.Header.MaxCombo;
+            scoreInfo.Statistics = frame.Header.Statistics;
+
+            Accuracy.Value = frame.Header.Accuracy;
+            Combo.Value = frame.Header.Combo;
+
+            scoreProcessor.ExtractScoringValues(frame.Header, out var currentScoringValues, out _);
+            TotalScore.Value = scoreProcessor.ComputeScore(Mode.Value, currentScoringValues, spectatorState.MaximumScoringValues);
+        }
+
+        protected override void Dispose(bool isDisposing)
+        {
+            base.Dispose(isDisposing);
+
+            // ReSharper disable once ConditionIsAlwaysTrueOrFalseAccordingToNullableAPIContract
+            if (spectatorClient != null)
+                spectatorClient.OnNewFrames -= onNewFrames;
```

Comment:
Should be able to write

```suggestion
            if (spectatorClient.IsNotNull())
                spectatorClient.OnNewFrames -= onNewFrames;
```

now.